File: Natixis Craftsmanship Practices/Breaking Dependencies/MarketTrading/csharp/UltimateTradingEngine.cs
using MarketTrading.External;

namespace MarketTrading;

public class UltimateTradingEngine
{
  public String ProcessOrder(string xmlOrder) {

        var mapper = new XmlObjectMapper();
        var sql = new NativeMarketSql();
        var http = new MarketHttpClient();
        var chaos = new MarketChaos();

        var order = mapper.Read<OrderXml>(xmlOrder, typeof(OrderXml));

        var exec = new ExecutionXml();
        exec.Executed = false;
        exec.Status = "INIT";
        exec.AvgPrice = 0.0;

        if (!MarketContext.MarketOpen) {
            exec.Status = "MARKET_CLOSED";
            return mapper.Write(exec);
        }

        MarketContext.ExecutionsToday++;

        var riskScore = 0;

        if (order.Qty > 1000) {
            riskScore += 2;
        }

        if ("SELL".Equals(order.Side)) {
            riskScore++;
        }

        if ((order.Flags & 1) == 1) {
            riskScore--;
        }

        var marketStatus = http.GetStatus(
                "http://exchange/api/status?symbol=" + order.Symbol
        );

        if (marketStatus != 200) {
            exec.Status = "MARKET_DOWN";
            return mapper.Write(exec);
        }

        var positionXml = sql.LoadPositionAsXml(order.Trader, order.Symbol);

        // BUG: position never really parsed
        if (positionXml.Contains("<qty>")) {
            if ("SELL".Equals(order.Side) && order.Qty > 500) {
                riskScore += 3;
            }
        }

        var now = chaos.Now();
        if (now % 2 == 0) {
            riskScore += chaos.Rnd(3);
        }

        var price = order.Price;

        if (MarketContext.LastPrices.ContainsKey(order.Symbol)) {
            var last = MarketContext.LastPrices[order.Symbol];
            if (Math.Abs(last - price) > 2) {
                price = (last + price) / 2;
            }
        }

        if (riskScore > 3) {
            exec.Status = "REJECTED_RISK";
        } else {
            double execPrice;
            if (chaos.Rnd(10) > 7) {
                execPrice = price + 0.1;
            } else {
                execPrice = price - 0.1;
            }

            exec.Executed = true;
            exec.ExecutionId = order.Id + "-" + MarketContext.ExecutionsToday;
            exec.AvgPrice = execPrice;
            exec.Status = "EXECUTED";

            MarketContext.LastPrices[order.Symbol] = execPrice;

            var tradeXml =
                    "<trade>" +
                            "<id>" + exec.ExecutionId + "</id>" +
                            "<symbol>" + order.Symbol + "</symbol>" +
                            "<qty>" + order.Qty + "</qty>" +
                            "<price>" + execPrice + "</price>" +
                            "</trade>";

            sql.PersistTradeAsXml(tradeXml);
        }

        return mapper.Write(exec);
    }
}

File: Natixis Craftsmanship Practices/Breaking Dependencies/MarketTrading/cpp/src/UltimateTradingEngine.cpp

#include "UltimateTradingEngine.h"

#include <string>

#include "MarketChaos.h"
#include "MarketHttpClient.h"
#include "NativeMarketSql.h"
#include "XmlParser.h"

std::string UltimateTradingEngine::processOrder(const std::string &xmlOrder) {

    NativeMarketSql sql;
    MarketHttpClient http;
    MarketChaos chaos;

    XmlParser<OrderXml> mapperOrder;
    OrderXml order = mapperOrder.read(xmlOrder);

    ExecutionXml exec;
    exec.executed = false;
    exec.status = "INIT";
    exec.avgPrice = 0.0;

    if (!MarketContext::marketOpen) {
        exec.status = "MARKET_CLOSED";
        XmlParser<ExecutionXml> mapperExecution;
        return mapperExecution.write(exec);
    }

    MarketContext::executionsToday++;

    int riskScore = 0;

    if (order.qty > 1000) riskScore += 2;
    if (order.side == "SELL") riskScore++;
    if ((order.flags & 1) == 1) riskScore--;

    int marketStatus = http.getStatus(
        "http://exchange/api/status?symbol=" + order.symbol
    );

    if (marketStatus != 200) {
        exec.status = "MARKET_DOWN";
        XmlParser<ExecutionXml> mapperExecution;
        return mapperExecution.write(exec);

    }

    std::string positionXml =
            sql.loadPositionAsXml(order.trader, order.symbol);

    if (!positionXml.empty() &&
        positionXml.find("<qty>") != std::string::npos) {
        if (order.side == "SELL" && order.qty > 500) {
            riskScore += 3;
        }
    }

    long now = chaos.now();
    if (now % 2 == 0) {
        riskScore += chaos.rnd(3);
    }

    double price = order.price;

    auto it = MarketContext::lastPrices.find(order.symbol);
    if (it != MarketContext::lastPrices.end()) {
        double last = it->second;
        if (std::abs(last - price) > 2.0) {
            price = (last + price) / 2.0;
        }
    }

    if (riskScore > 3) {
        exec.status = "REJECTED_RISK";
    } else {
        double execPrice =
                (chaos.rnd(10) > 7) ? price + 0.1 : price - 0.1;

        exec.executed = true;
        exec.executionId =
                order.id + "-" + std::to_string(MarketContext::executionsToday);
        exec.avgPrice = execPrice;
        exec.status = "EXECUTED";

        MarketContext::lastPrices[order.symbol] = execPrice;

        std::string tradeXml =
                "<trade>"
                "<id>" + exec.executionId + "</id>"
                "<symbol>" + order.symbol + "</symbol>"
                "<qty>" + std::to_string(order.qty) + "</qty>"
                "<price>" + std::to_string(execPrice) + "</price>"
                "</trade>";

        sql.persistTradeAsXml(tradeXml);
    }

    XmlParser<ExecutionXml> mapperExecution;
    return mapperExecution.write(exec);
}